sopel/modules/ip.py

```diff
@@ -112,80 +287,246 @@ def _find_geoip_db(bot):
+    if 'WHOX' in bot.isupport:
+        # WHOX syntax, see http://faerion.sourceforge.net/doc/irc/whox.var
+        # Needed for accounts in WHO replies. The random integer is a param
+        # to identify the reply as one from this command, because if someone
+        # else sent it, we have no way to know what the format is.
+
+        # 'x' indicates uncloaked address. This is triggered by
+        # RPL_YOUREOPER because that functionality is restricted to opers.
+        rand = str(randint(0, 999))
+        while rand in who_reqs:
+            rand = str(randint(0, 999))
+        LOGGER.debug('Sending who: %s', rand)
+        who_reqs[rand] = True
+        bot.write([r'WHO * n%nuhti,' + rand])
+
+#:safe.oh.us.irc.scpwiki.com NOTICE Kufat :*** CONNECT: Client connecting on port 6697 (class main): ASNbot!sopel@ool-45734b07.dyn.optonline.net (69.115.75.7) [Sopel: https://sopel.chat/]
+@module.rule(r'.*Client connecting .*: (\S*)!(\S*)@(\S*) \((.*)\)')
+@module.event("NOTICE")
+@module.priority("high")
+def handle_snotice_conn(bot, trigger):
+    LOGGER.debug("Saw connect line: [%s] from [%s]", trigger.raw, trigger.sender)
+    #Only servers may have '.' in the sender name, so this isn't spoofable
+    if "scpwiki.com" in trigger.sender:
+        nick, user, host, ip = trigger.groups()
+        examine_user(bot, user, ip, None, nick) #cloaked host not known
+        #Be **certain** we don't waste our lookups on irccloud
+        if any(host.endswith(s) for s in (".irccloud.com", ".mibbit.com")):
+            return
+        # We need to check if the IP is in any exempt CIDR ranges
+        if get_exemption(ip):
+            return
+
+        res = examine_user(bot, user, ip, host, nick)
+        if res:
+            # Acted on above; just log here
+            score, is_proxy, is_recent_abuse = res
+            LOGGER.debug(f"handle_snotice_conn: {nick}!{user}@{host} ({ip}) had "
+                          "score {score}, proxy: {is_prox}, "
+                          "recent_abuse: {is_recent_abuse}")
+
+# NICK: User Kufat-bar changed their nickname to Kufat-foo
+# This is redundant for users the bot can see in-channel but needed for users with no common channel
+@module.rule(r'.*User (\S+) changed their nickname to (\S+).*')
+@module.event("NOTICE")
+@module.priority("high")
+def handle_snotice_ren(bot, trigger):
+    LOGGER.debug("Saw nick change line: [%s] from [%s]", trigger.raw, trigger.sender)
+    if "scpwiki.com" in trigger.sender:
+        oldnick = Identifier(trigger.group(1))
+        newnick = Identifier(trigger.group(2))
+        if olduser := bot.users.get(oldnick):
+            populate_user(bot, olduser.user, olduser.ip, olduser.host, newnick)
+
+@module.require_privilege(module.OP)
+@module.commands('ip_exempt')
+@module.example('.ip_exempt 8.8.8.8')
+def ip_exempt(bot, trigger):
+    if not trigger.group(3):
+        return bot.reply("You must specify an IP or range in CIDR format to exempt.")
+    elif not trigger.group(4):
```

Comment:
Where are these capturing groups specified? Group 4 is mandated but appears to be unused

Reply:
Yeah, it's a weird sopel thing. Group 4 is command-line #2; if there's no second argument, group 4 is none. But when it comes time to actually _use_ the description string, then we want the whole thing so we need to take group 2, which is all arguments together, strip off the first arg, and use that. It's taken from existing sopel code that does something similar (IIRC it was .tell)